### PR TITLE
Added RGBA5551 compatibility in the rasterizer

### PR DESCRIPTION
**Pull request by archshift**
This allows Virtual Console games to display properly.

### CODE REVIEW

**Comment by archshift** [comment]
Note: this depends on #599.

**Comment by archshift** [comment]
Rebased onto master.

File: src/video_core/rasterizer.cpp [hunk 1]
```diff
@@ -51,6 +51,16 @@ static void DrawPixel(int x, int y, const Math::Vec4<u8>& color) {
         break;
     }
 
+    case registers.framebuffer.RGBA5551:
+    {
+        u16* pixel = (u16*)(color_buffer + dst_offset);
+        *pixel = (Color::Convert8To5(color.r()) << 11)   |
+                    (Color::Convert8To5(color.g()) << 6) |
+                    (Color::Convert8To5(color.b()) << 1) |
+                    Color::Convert8To1(color.a());
```
Comment (bunnei) [inline src/video_core/rasterizer.cpp]: Can you indent these lines so that they all line up?

**Comment by bunnei** [comment]
Mostly LGTM

File: src/video_core/rasterizer.cpp (hunk 1 above)
Comment (archshift) [inline src/video_core/rasterizer.cpp]: But they do all line up :P

File: src/video_core/color.h [hunk 2]
```diff
@@ -28,5 +28,14 @@ static inline u8 Convert6To8(u8 value) {
     return (value << 2) | (value >> 4);
 }
 
+/// Convert a 8-bit color component to 5 bit
+static inline u8 Convert8To5(u8 value) {
+    return value * 0x1F / 0xFF;
```
Comment (neobrain) [inline src/video_core/color.h]: Use `>> 3` instead for consistency with the other functions.

File: src/video_core/color.h [hunk 3]
```diff
@@ -28,5 +28,14 @@ static inline u8 Convert6To8(u8 value) {
...
+}
+
+/// Convert a 8-bit color component to 1 bit
+static inline u8 Convert8To1(u8 value) {
+    return value > (0xFF / 2) ? 1 : 0;
```
Comment (neobrain) [inline src/video_core/color.h]: `>> 7` here.

File: src/video_core/rasterizer.cpp [hunk 4]
```diff
@@ -51,6 +51,16 @@ static void DrawPixel(int x, int y, const Math::Vec4<u8>& color) {
         break;
     }
 
+    case registers.framebuffer.RGBA5551:
+    {
+        u16* pixel = (u16*)(color_buffer + dst_offset);
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: This is not big-endian compatible. Please fill the individual bytes instead.

File: src/video_core/rasterizer.cpp [hunk 5]
```diff
@@ -90,6 +100,17 @@ static const Math::Vec4<u8> GetPixel(int x, int y) {
         return ret;
     }
 
+    case registers.framebuffer.RGBA5551:
+    {
+        Math::Vec4<u8> ret;
+        u16 pixel = *(u16*)(color_buffer + src_offset);
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Again, not big-endian compatible.

File: src/video_core/rasterizer.cpp [hunk 6]
```diff
@@ -90,6 +100,17 @@ static const Math::Vec4<u8> GetPixel(int x, int y) {
         return ret;
     }
 
+    case registers.framebuffer.RGBA5551:
+    {
+        Math::Vec4<u8> ret;
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: There's actually not much of a reason to redeclare this for each switch case, is there?

File: src/video_core/rasterizer.cpp (hunk 4 above)
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: (Also, similarly to below, this needn't be redeclared for each switch case)

**Comment by neobrain** [comment]
LGTM once big-endian compatibility and consistency issues are addressed.

**Comment by neobrain** [comment]
Actually, repost from #617, which apparently is meant to be merged after this PR:

Since it's now more relevant than ever, We should explain something like this explicitly:

// Components are laid out in reverse byte order, most significant bits first.

Could you add this comment to pica.h (ColorFormat) and replace the one in gpu.h (PixelFormat) with something like this?

File: src/video_core/color.h (hunk 2 above)
Comment (archshift) [inline src/video_core/color.h]: That would change the meaning of this code, though.

File: src/video_core/color.h (hunk 3 above)
Comment (archshift) [inline src/video_core/color.h]: Again, that would change the meaning of the code.

**Comment by archshift** [comment]
Should be good to go.

File: src/video_core/color.h [hunk 7]
```diff
@@ -28,5 +28,24 @@ static inline u8 Convert6To8(u8 value) {
...
 }
 
+/// Convert a 8-bit color component to 1 bit
+static inline u8 Convert8To1(u8 value) {
+    return value >> 7;
+}
+
+/// Convert a 8-bit color component to 4 bit
+static inline u8 Convert8To4(u8 value) {
+    return value >> 7;
```
Comment (neobrain) [inline src/video_core/color.h]: typo here?

Comment (archshift) [inline src/video_core/color.h]: Indeed.

**Comment by yuriks** [comment]
LGTM

**Comment by neobrain** [comment]
LGTM

**Pull request merged**